pymc/smc/smc.py

```diff
@@ -570,3 +547,86 @@ def _logp_forw(point, out_vars, in_vars, shared):
     f = compile_pymc([inarray0], out_list[0])
     f.trust_input = True
     return f
+
+
+class Particles:
+    """
+    List of draws from variables (particles)
+    belonging to a model that can be modified
+    in place as if they were an array.
+    """
+
+    @classmethod
+    def build(cls, draws: int, initial_point, init_rnd, model, dual_representation=False):
+        var_info = {}
+        variables = inputvars(model.value_vars)
+        for v in variables:
+            var_info[v.name] = (initial_point[v.name].shape, initial_point[v.name].size)
+        points = [
+            Point({v.name: init_rnd[v.name][i] for v in variables}, model=model)
+            for i in range(draws)
+        ]
+        draws = draws
+
+        return Particles(points, draws, var_info, variables, model)
+
+    def __init__(self, points, draws, var_info, variables, model):
+        self.as_array = np.array(floatX([DictToArrayBijection.map(point).data for point in points]))
+        self.variables = variables
+        self._draws = draws
+        self._var_info = var_info
+        self._model = model
+
+    def resample(self, resampling_indexes):
+        self.as_array = self.as_array[resampling_indexes]
+
+    def __len__(self):
+        return self._draws
+
+    def __getitem__(self, item):
+        return self.as_array[item]
+
+    def set(self, indexes, values):
+        """
+        Consistently updates array and points
+        representations. If no index is selected, modify all
+        """
+        if indexes is None:
+            if not len(values) == len(self):
+                raise ValueError(
+                    "If values size doesn't match particle size, then indexes need to be specified"
+                )
+            indexes = [True] * len(self)
```

Comment:
Why not just slice?  Creating a list seems wasteful

Reply:
Removed the index construction altogether